### PR TITLE
[AP] cleanup specfiles and sidebar

### PR DESCRIPTION
**Pull request by ElliotFriend**
## What's this, then?

To get the site ready for us to add docs versioning to AP (#345), there is some preparation that should be done. First, was upgrading to docusaurus v3, and now it's cleaning up a bit how we have docusaurus handling the AP files. This PR aims to tackle that second part.

~~We're consolidates the callback, platform, and custody specfiles into a single `main.yaml` file (notwithstanding the external schemas, etc. files).~~ (Edit: We're not consolidating, but still doing some cleanup) This also upgrades the SDP specfile to openapi v3.1, and fixes some spec errors in the file.

## Remains to be done:

- [x] reconfigure build to break on broken links
- [ ] add some nginx redirects since some URL paths get shifted up a directory

## TL;DR

There's a lot happening in this PR, so here's a recap of what I've done:

- consolidated the three distinct API specifications into a single specfile
- made the rpc spec behave/generate more like the soroban rpc spec (including the documentation pages generated from it)
- fiddled with the sidebar definition so we don't need to custom-code a generator
- because the custom sidebar generator is no more, i had to make some sidebar adjustments to the SDP docs, too. no big changes there, though.
- other bits-n-bobs related to the redocly cli (updates, configs, etc.)

### CODE REVIEW

File: openapi/anchor-platform/main.yaml [hunk 1]
```diff
@@ -1,26 +1,241 @@
-openapi: 3.0.0
+openapi: 3.1.0
```
Comment (Ifropc) [inline openapi/anchor-platform/main.yaml:1]: Do we need to store all openapi stuff in `main.yaml` from now on?

File: openapi/anchor-platform/main.yaml [hunk 2]
```diff
@@ -1,26 +1,241 @@
...
-    anchor's responses to these requests affect the Platform responses to the client application.
-  title: Synchronous Callbacks API
+    The platform also includes features for managing assets, transactions, and user accounts, and supports a variety of deployment options, including using Docker or Kubernetes. Overall, the Anchor Platform aims to simplify the process of building and managing a Stellar-based financial service, allowing businesses to focus on providing value to their customers.
+
+    Learn about integrating with the Anchor Platform in the [Admin Guide](https://developers.stellar.org/platforms/anchor-platform/admin-guide) or get API information in the [API Reference](https://developers.stellar.org/platforms/anchor-platform/api-reference).
+  termsOfService: https://stellar.org/terms-of-service
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
 tags:
```
Comment (Ifropc) [inline openapi/anchor-platform/main.yaml:16]: Everything below is copy of Platform API + Custody API right?

**Review by Ifropc — COMMENTED** [review]
Do you mind adding a small `README` in `/openrpc` folder to maybe add some information on how to add new method? (I think changing existing methods is relatively straight-forward but if you think it should also be documented it'd be a nice addition as well)

File: patches/@metamask+open-rpc-docs-react+0.1.2.patch [hunk 3]
```diff
@@ -21,12 +21,15 @@ index 354ce73..129371f 100644
 +            result: examplePairing.result.value
 +        }
 +        // const jsCode = `await window.ethereum.request(${JSON.stringify(methodCall, null, "  ")});`;
++        const serverUrl = methodName.indexOf('_') > -1
```
Comment (Ifropc) [inline patches/@metamask+open-rpc-docs-react+0.1.2.patch:24]: Not sure if method name containing `_` is a good enough way to distinguish it... We may add something on either RPC side that could break this logic. Is there anything else we can use? Tags maybe?

File: docs/data/rpc/api-reference/methods/getEvents.mdx [hunk 4]
```diff
@@ -4,5 +4,8 @@ description: Returns contract events
 ---
 
 import { RpcMethod } from "@site/src/components/RpcMethod";
+import rpcSpec from "@site/static/stellar-rpc.openrpc.json";
 
-<RpcMethod method="getEvents" platform="soroban" />
+<RpcMethod
+  method={rpcSpec.methods.filter((meth) => meth.name === "getEvents")[0]}
```
Comment (Ifropc) [inline docs/data/rpc/api-reference/methods/getEvents.mdx:10]: Could it be refactored so it's possible to call in like before?
`<RpcMethod method="getEvents">`

**Review by Ifropc — APPROVED** [review]

File: patches/@metamask+open-rpc-docs-react+0.1.2.patch (hunk 3 above)
Comment (ElliotFriend) [reply patches/@metamask+open-rpc-docs-react+0.1.2.patch:24]: You're absolutely right, this is a **really fragile** way to make this check. We're well on our way to publishing [our own fork](https://github.com/stellar/open-rpc-docs-react) of the package, and this patch file will (thankfully) be out of the picture soon. So, it seemed like this quick-n-dirty bandaid might be sufficient just for the time-being.

We could definitely do a more robust check now, if you'd prefer, though. iirc the stellar-rpc doesn't use tags for _any_ of its methods, so that probably is a better way to go about it.

File: docs/data/rpc/api-reference/methods/getEvents.mdx (hunk 4 above)
Comment (ElliotFriend) [reply docs/data/rpc/api-reference/methods/getEvents.mdx:10]: The short answer is "I'm not sure if we can."

The reason for co-locating the specfile _into_ the `/platforms/anchor-platform` directory is to make it a part of the versioned docs so when someone clicks from v2 to v3, the proper specfile gets rendered and displayed.

But, the `RpcMethod` React component lives entirely outside of that context. I'm not sure how to make the component work across differing versions (or non-versioned like stellar-rpc) while still passing in `getEvent`-like strings like we had before (which i do agree has much better ergonomics). I'm open to ideas or suggestions, for sure, I was just drawing a blank.

File: openapi/anchor-platform/main.yaml (hunk 1 above)
Comment (ElliotFriend) [reply openapi/anchor-platform/main.yaml:1]: At the moment, yes. That's the "heart" of the specfile, kindof like the "index.html" page. In the future, we could certainly break it into a more manageable series of smaller files (similar to how Horizon is done (but also done better, that ones a bit of a rat's nest lol)) to get bundled together (like how `yarn api` bundles them already).

Comment (ElliotFriend) [reply openapi/anchor-platform/main.yaml:1]: Even if we did break things into sub-files, they will still need to get `$ref`ed in the `main.yaml` file in order to be bundled up.

File: openapi/anchor-platform/main.yaml (hunk 2 above)
Comment (ElliotFriend) [reply openapi/anchor-platform/main.yaml:16]: correct

Comment (ElliotFriend) [reply openapi/anchor-platform/main.yaml:16]: Actually, Platform API + Custody API + Callbacks API

File: patches/@metamask+open-rpc-docs-react+0.1.2.patch (hunk 3 above)
Comment (Ifropc) [reply patches/@metamask+open-rpc-docs-react+0.1.2.patch:24]: Sounds good, if it's temporary let's just leave the glue there for now

File: docs/data/rpc/api-reference/methods/getEvents.mdx (hunk 4 above)
Comment (Ifropc) [reply docs/data/rpc/api-reference/methods/getEvents.mdx:10]: > to make it a part of the versioned docs so when someone clicks from v2 to v3, the proper specfile gets rendered and displayed.

Got it, I didn't think about it, makes sense! 
Not sure if it can be done in a better manner then 😔

File: openapi/anchor-platform/main.yaml (hunk 2 above)
Comment (Ifropc) [reply openapi/anchor-platform/main.yaml:16]: Got it, any reason for merging them back together? I liked the previous one because it's easy to navigate, but don't have any strong preference. 
Also cc @philipliu for opinion on merging this yaml files together

File: openapi/anchor-platform/main.yaml (hunk 1 above)
Comment (Ifropc) [reply openapi/anchor-platform/main.yaml:1]: Yep I think that was how it was done before, right?

File: openrpc/README.md [hunk 5]
```diff
@@ -13,17 +14,58 @@
...
+If you're here to add a new method to one of the specfiles, you'll want to:
+
+1. Create a new `<method_name>.json` file in the relevant
+   `/openrpc/src/{stellar-rpc,anchor-platform}/methods` directory. Follow the
+   OpenRPC specification concerning [method objects] for defining this file.
+2. Feel free to use any existing (or create new) `$ref` objects along the way.
+   More about them [later on](#json-refs).
+3. Once your method is defined the way you want it, run the [build](#building)
+   and [validate](#testing) scripts to generate updated "bundled" specfiles.
+4. Create a new Docusaurus page for the method. You can use one of the existing
```
Comment (Ifropc) [inline openrpc/README.md:28]: Perfect thank you very much! I honestly spent waaay too long to figure out how to do that when I needed to 😓

File: openapi/anchor-platform/main.yaml (hunk 2 above)
Comment (philipliu) [reply openapi/anchor-platform/main.yaml:16]: I would prefer if we could keep them separate

**Comment by Ifropc** [comment]
LGTM, thanks @ElliotFriend ! We can always get back to splitting API files later on

**Review by Ifropc — APPROVED** [review]

File: openapi/anchor-platform/main.yaml (hunk 2 above)
Comment (ElliotFriend) [reply openapi/anchor-platform/main.yaml:16]: The reasoning was that it's a bit of a "pre-emptive" move for the versioning setup. It would be much simpler to coordinate and manage the versions of a single specification file, rather than 3 or 4 different files.

That said, my mind has been moving in the same direction as you both over the last couple days. I think  there are a couple options to address it. I'm gonna dig a bit into the [versioning capabilities](https://docusaurus-openapi.tryingpan.dev/versioning) of the openapi plugin we're using, as well as this [experimental `join` command](https://redocly.com/docs/cli/commands/join) in the Redocly CLI. I think both could provide a reasonable option for what we're after (which is: A useful, maintainable specfile (or set of them) that cleanly and nicely generates the documentation pages for us).

**Comment by ElliotFriend** [comment]
Updates since last major review:

- We've gone back to keeping the various specs of the Anchor Platform as different files. file management in the repo is a _little_ more cumbersome, but the trade-offs are worth it.
- The sidebar is generated now in a much simpler manner, it's just an autogenerated directory 🎉 
- The `redocly` CLI is configured to include and manage commands for the different parts of the AP specs
- The sidebar is now "flattened" out a bit more, so there are fewer dropdowns that need to be navigated through for users
- I've moved the RPC spec documentation _into_ the "Platform Server" category, since the methods exist on the same instance/server. (we can easily undo this, it felt like it made some sense, though?)

Here's the PR Preview:
https://developers-pr872.previews.kube001.services.stellar-ops.com/platforms/anchor-platform

Would love another review to see how we think this'll work out.

**Review by Ifropc — APPROVED** [review]
